### PR TITLE
[expo-analytics-amplitude] Add `setTrackingOptions` so that Amplitude doesn't track PII

### PR DESCRIPTION
**Pull request by amaury1093**
# Why

➡️ Let apps using Amplitude be less evil.

The current Amplitude in Expo only exposes a subset of all the functions available in the iOS/Android SDKs of Amplitude.

One important feature missing is the ability to disable tracking of certain user properties, such as [PII](https://en.wikipedia.org/wiki/Personal_data). Right now, all apps using Amplitude (including the Expo app itself!) track the IP address, exact lat/lng, IDFA/ADID of every user 😕...

As an app developer using Expo and Amplitude, I'd like to have the choice to not track PII.

# How

I followed Amplitude's docs:
- android: https://help.amplitude.com/hc/en-us/articles/115002935588-Android-SDK-Installation#disable-automatic-tracking-of-user-properties
- ios: https://help.amplitude.com/hc/en-us/articles/115002278527-iOS-SDK-Installation#disable-automatic-tracking-of-user-properties

# Test Plan

tbh I'm not sure how to test this, if anyone has an idea it would be great.

# Note

I've done some JS/RN dev on top of the Expo framework, but I'm very new to iOS & Android native development, and would like to use [Hacktoberfest](https://hacktoberfest.digitalocean.com/) to learn it. That's why I would greatly appreciate any help on this PR 🚀 .

### CODE REVIEW

**Comment by brentvatne** [comment]
this looks pretty reasonable to me! @EvanBacon any tips on how to test this?

File: packages/expo-analytics-amplitude/ios/EXAmplitude/EXAmplitude.m
Comment (sjchmiela) [inline packages/expo-analytics-amplitude/ios/EXAmplitude/EXAmplitude.m]: I think if one would provide an object like
```js
{
  "disableCarrier": false
}
```
what `NSDictionary *options` would represent would be
```objc
{
  @"disableCarrier": [NSNumber numberWithBool:false]
}
```
so the value would be a non-null pointer to a number representing a false (0) value. `if`s on such values are always positive (pointer will be `> 0`).

What I would suggest would be to:
1. change notation to a simpler `options[@"disableCarrier"]` (a cosmetic change, with regex shouldn't be that difficult and it makes the code more consistent)
2. instead of doing `if (dictionaryValue)` let's do `if ([dictionaryValue boolValue])`. If the value will be empty, it will evaluate to `false` (developer hasn't specified `disableCarrier: anything`, we don't want to disable carrier), if the value will be a number (or a boolean) it will evaluate properly and otherwise it will reject the promise with a native error (appropriate behavior).

So all in all:
```objc
if ([options[@"disableCarrier"] boolValue]) {
```

File: packages/expo-analytics-amplitude/ios/EXAmplitude/EXAmplitude.m
Comment (sjchmiela) [inline packages/expo-analytics-amplitude/ios/EXAmplitude/EXAmplitude.m]: Is it possible that `initializeApiKey:` triggers an event to Amplitude? What my question suggests is: should we add a `trackingOptions` argument to `initialize` method?

File: packages/jest-expo/src/preset/expoModules.js
Comment (sjchmiela) [inline packages/jest-expo/src/preset/expoModules.js]: Good job! 👏

File: packages/expo-analytics-amplitude/android/src/main/java/expo/modules/analytics/amplitude/AmplitudeModule.java
Comment (sjchmiela) [inline packages/expo-analytics-amplitude/android/src/main/java/expo/modules/analytics/amplitude/AmplitudeModule.java]:
```suggestion
  public void setTrackingOptions(final ReadableArguments options, Promise promise) {
```

[`ReadableArguments`](https://github.com/expo/expo/blob/master/packages/@unimodules/core/android/src/main/java/org/unimodules/core/arguments/ReadableArguments.java) is a unimodules class similar to RN's `ReadableMap` which allows calls like `options.getBoolean('disableAdid', false)` which properly returns a boolean, not a pointer to an object.

**Review by sjchmiela — COMMENTED** [review]
Thank you for the contribution, this is a nice high quality, readable PR!

Could you please look through the comments and see if you could fix the issues? Would you also care to add a CHANGELOG entry? 🙂

File: packages/expo-analytics-amplitude/ios/EXAmplitude/EXAmplitude.m
Comment (amaurymartiny) [reply packages/expo-analytics-amplitude/ios/EXAmplitude/EXAmplitude.m]: My 2 arguments why I personally prefer a separate `setTrackingOptions`:
- it's a one-to-one mapping with Amplitude's SDK
- the developer might disable some tracking on init (like IP address, lat/lng), but after the user clicks on "I accept the terms", enable them. So we anyways need a separate `setTrackingOptions` function for this use case.

**Comment by amaurymartiny** [comment]
The CI errors seem unrelated, how can i fix them?

**Comment by sjchmiela** [comment]
CI doesn't pass for external contributors 🤷‍♂ security matter

**Comment by sjchmiela** [comment]
(We sign the app with Expo keys, so we expose real password etc. to build job)

File: packages/expo-analytics-amplitude/android/src/main/java/expo/modules/analytics/amplitude/AmplitudeModule.java
Comment (sjchmiela) [inline packages/expo-analytics-amplitude/android/src/main/java/expo/modules/analytics/amplitude/AmplitudeModule.java]: A comment on https://github.com/expo/expo/pull/5877/files#r334839075

Here, instead of rejecting unless initialized we could save the options and apply in between

```
mClient = new AmplitudeClient();
< here > 
mClient.initialize(getContext(), apiKey);
promise.resolve(null);
```
so that tracking options are applied before the client is initialized (with your implementation it is possible on iOS, just not on Android). Refactoring lines 100-146 to a separate function, remembering last set tracking options before initialization and applying options before setting API key should do the work. 🤔

Comment (amaurymartiny) [reply packages/expo-analytics-amplitude/android/src/main/java/expo/modules/analytics/amplitude/AmplitudeModule.java]: Good idea. https://github.com/expo/expo/pull/5877/commits/94d75cdbbb61f92f912025c5ce7b034b577f5cb8 should fix this

**Comment by wodin** [comment]
Random thought about how one might test this. This will not work with key pinning, but it seems that [Amplitude discourages that](https://help.amplitude.com/hc/en-us/articles/115002935588-Android-SDK-Installation#ssl-pinning) anyway:

- Figure out what host(s) is(are) contacted when sending this Amplitude data.
- Generate an SSL CA certificate.
- Set up an HTTP proxy server using a cert with subjectAlternativeName containing the proxy's IP address and signed by the above CA cert.
- Install the CA cert onto an Android device. I think I've done this in the past, but I might be misremembering. Not sure if this is possible on iOS, but I suspect it is.
- Set up a DNS server that responds with the proxy's IP address for requests to the Amplitude server(s) found in the first step.
- Log the traffic sent to the proxy before and after the change to see if it contains the tracked data.

@brentvatne is there something that should still be done with this PR before merging? e.g. the above convoluted testing? :)

**Review by sjchmiela — APPROVED** [review]
Thank you for bearing with us! I have put final touches on the PR (also noticed a big problem in `expo-analytics-amplitude` — unresolved promises), I'll wait for the CI to ensure `client_android` builds and it's ready to be merged!

**Pull request merged**
